### PR TITLE
Merge of development into main

### PR DESCRIPTION
**Pull request by caseyjlaw**

### CODE REVIEW

File: dsautils/dsa_store.py [hunk 1]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
 
-    def get_dict(self, key: "String") -> "Dictionary":
+    def _strict_json(self, val: str):
+        """Function will be called by json.loads with one of the following
+        strings: 'NaN', '-Infinity' or 'Infinity' for invalid numbers.
+
+        """
+        raise ValueError
+
+    def delete(self, key: str, dir_flag: bool = False, recursive: bool = False):
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:144]: approved function.

File: dsautils/dsa_store.py [hunk 2]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
+        :param dir_flag: Throws error if key is not empty
+        :param recursive: True to recursively delete keys
+        :type key: string
+        :type dir_flag: boolean
+        :type recursive: boolean
+
+        """
+        self.etcd.delete(key, dir_flag, recursive)
+    
+    def get_dict(self, key: str, parse_func: object = 'default') -> "Dictionary":
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:157]: approved. Signature change backwords compatible.

File: dsautils/dsa_store.py [hunk 3]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
         """Get data from Etcd store in the form of a dictionary for the
         specified key.
 
         :param key: Etcd key from which to read data.
+        "param parse_func: Set to None to allow NaN, Infinity and -Infinity
         :type key: String (Ex. '/mont/snap/1')
+        :type parse_func: Function which takes a string.
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:164]: approved doc strings.

File: dsautils/dsa_store.py [hunk 4]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
         """
 
         self.log.function('get_dict')
+        if parse_func == 'default':
+            parse_func = self._strict_json
+
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:170]: approved.

File: dsautils/dsa_store.py [hunk 5]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
+        if parse_func == 'default':
+            parse_func = self._strict_json
+            
         # etcd returns a 2-tuple. We want the first element
         data = self.etcd.get(key)[0]
         if data is not None:
             try:
-                return json.loads(data.decode("utf-8"))
+                return json.loads(data.decode("utf-8"),
+                                  parse_constant=parse_func)
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:176]: approved.

File: dsautils/dsa_store.py [hunk 6]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
             except:
                 self.log.error('could not convert json to dictionary')
                 raise
         else:
             self.log.warning('Nothing returned for key: {}'.format(key))
 
-    def add_watch_prefix(self, key: str, cb_func: "function") -> int:
+    def add_watch_prefix(self, key: str, cb_func: "function",
+                         parse_func: "function" = 'default') -> int:
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:184]: approve sig. change. bw compat.

File: dsautils/dsa_store.py [hunk 7]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
         callback to be called and 'dict' is the payload for that key.
+        parse_func if defined will be call with either 'Nan", '-Infinity' or
+        'Infinity' string type. Set to None to allow these values.
 
         :param key: Key prefix to watch. Callback function will be called when contents of any key starting with prefix changes.
         :param cb_func: Callback function. Must take list as argument.
+        :param parse_func: Set to None to allow NaN, -Infinity, Infinity
         :type key: str
         :type cb_func: function
+        :type parse_func: Function which takes a string.
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:199]: approve docstring changes.

File: dsautils/dsa_store.py [hunk 8]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
         :rtype: int The watch id for the callback. Can be used to cancel watch.
 
         """
 
-        watch_id = self.etcd.add_watch_prefix_callback(key, self._process_cb_prefix(cb_func))
+        if parse_func == 'default':
+            parse_func = self._strict_json
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:205]: duplicate code. turn into private function and replace other occurrences.

File: dsautils/dsa_store.py [hunk 9]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
+
+        watch_id = self.etcd.add_watch_prefix_callback(key,
+                                                       self._process_cb_prefix(cb_func, parse_func))
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:208]: approved.

File: dsautils/dsa_store.py [hunk 10]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
         self.watch_ids.append(watch_id)
         return watch_id
         
-    def add_watch(self, key: "String", cb_func: "Callback Function") -> int:
+    def add_watch(self, key: str, cb_func: "Callback Function",
+                  parse_func: "function" = 'default') -> int:
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:213]: approved sig.change. bw compat.

File: dsautils/dsa_store.py [hunk 11]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
         dictionary will represent the payload associated with the key.
+        parse_func if defined will be call with either 'Nan", '-Infinity' or
+        'Infinity' string type. Set to None to allow these values.
 
         :param key: Key to watch. Callback function will be called when contents of key changes.
         :param cb_func: Callback function. Must take dictionary as argument.
+        :param parse_func: Set to None to allow NaN, -Infinity, Infinity
         :type key: String
         :type cb_func: Function(dictionary)
+        :type parse_func: Function which takes a string.
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:226]: approve docstring additions.

File: dsautils/dsa_store.py [hunk 12]
```diff
@@ -107,78 +110,128 @@ def get_etcd(self) -> "Etcd3Client Object":
...
         :rtype: int
 
         """
 
-        watch_id = self.etcd.add_watch_callback(key, self._process_cb(cb_func))
+        if parse_func == 'default':
+            parse_func = self._strict_json
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:232]: duplicate code. replace with function.

File: dsautils/dsa_store.py [hunk 13]
```diff
@@ -196,16 +249,22 @@ def get_watch_ids(self) -> "List":
         """
         return self.watch_ids
 
-    def _process_cb(self, cb_func: "Callback Function"):
+    def _process_cb(self, cb_func: "Callback Function",
+                    parse_func: "function" = 'default'):
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:253]: approve sig. change. bw compat.

File: dsautils/dsa_store.py [hunk 14]
```diff
@@ -196,16 +249,22 @@ def get_watch_ids(self) -> "List":
...
         """Private closure to call callback function with dictionary argument
         representing the payload of the key being watched.
 
         :param cb_func: Callback function. Takes a dictionary argument.
+        :param parse_func: Set to None to allow NaN, -Infinity, Infinity
         :type cb_func: Function
+        :type parse_func: Function which takes a string.
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:260]: approve docstring additions.

File: dsautils/dsa_store.py [hunk 15]
```diff
@@ -196,16 +249,22 @@ def get_watch_ids(self) -> "List":
...
         """
 
         self.log.function('_process_cb')
 
+        if parse_func == 'default':
+            parse_func = self._strict_json
+
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:267]: duplicate code. replace with function.

File: dsautils/dsa_store.py [hunk 16]
```diff
@@ -221,7 +280,7 @@ def a(event):
                         value = ev.value.decode('utf-8')
                         # parse the JSON command into a dict.
                         try:
-                            payload = self._parse_value(value)
+                            payload = self._parse_value(value, parse_func)
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:283]: approved.

File: dsautils/dsa_store.py [hunk 17]
```diff
@@ -236,16 +295,22 @@ def a(event):
                 raise
         return a
 
-    def _process_cb_prefix(self, cb_func: "Callback Function"):
+    def _process_cb_prefix(self, cb_func: "Callback Function",
+                           parse_func: "function" = 'default'):
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:299]: approved. private func. sig change.

File: dsautils/dsa_store.py [hunk 18]
```diff
@@ -236,16 +295,22 @@ def a(event):
...
         """Private closure to call callback function with list argument
         representing the kay and payload of the keys being watched.
 
         :param cb_func: Callback function. Takes a list argument.
+        :param parse_func: Set to None to allow NaN, -Infinity, Infinity
         :type cb_func: Function
+        :type parse_func: Function which takes a string.
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:306]: approved docstring additions.

File: dsautils/dsa_store.py [hunk 19]
```diff
@@ -236,16 +295,22 @@ def a(event):
...
         """
 
         self.log.function('_process_cb')
 
+        if parse_func == 'default':
+            parse_func = self._strict_json
+
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:313]: duplicate code. replace with function.

File: dsautils/dsa_store.py [hunk 20]
```diff
@@ -277,21 +342,29 @@ def a(event):
                 raise
         return a
 
-    def _parse_value(self, value: "Json String") -> "Dictionary":
+    def _parse_value(self, value: "Json String",
+                     parse_func: "function" = 'default') -> "Dictionary":
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:346]: approved.

File: dsautils/dsa_store.py [hunk 21]
```diff
@@ -277,21 +342,29 @@ def a(event):
...
         """Parse the string in JSON format into a dictionary.
 
         :param value: JSON string of the form: {"key":"value"}
                       or {"key":number|bool}
+        "param parse_func: Set to None to allow NaN, Infinity and -Infinity
         :type value: String
+        :type key: String (Ex. '/mont/snap/1')
+        :type parse_func: Function which takes a string.
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:354]: approved docstring

File: dsautils/dsa_store.py [hunk 22]
```diff
@@ -277,21 +342,29 @@ def a(event):
...
         :return: Key,value dictionary
         :rtype: Dictionary
         :raise: ValueError
         """
 
         self.log.function('_parse_value')
+
+        if parse_func == 'default':
+            parse_func = self._strict_json
+
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:364]: duplicate code. replace with function.

File: dsautils/dsa_store.py [hunk 23]
```diff
@@ -277,21 +342,29 @@ def a(event):
...
         rtn = {}
         try:
-            rtn = json.loads(value)
+            rtn = json.loads(value, parse_constant=parse_func)
```
Comment (rh-codebase) [inline dsautils/dsa_store.py:367]: approved.

File: test/test_dsa_store.py [hunk 24]
```diff
@@ -33,6 +34,98 @@ def test_put_get(self):
         rtn_dict = my_etcd.get_dict('/test/1')
         self.assertEqual(rtn_dict, test_dict)
 
+    def put_bad_val(self, val):
```
Comment (rh-codebase) [inline test/test_dsa_store.py:37]: general comment. If these test keys are being placed in the production etcd service, then they should be deleted when the test is complete.

File: test/test_dsa_store.py [hunk 25]
```diff
@@ -33,6 +34,98 @@ def test_put_get(self):
...
+
+    def test_get_Nan_allow(self):
+        self._get_val_allow(math.nan)
+
+    def test_get_minus_Infinity_allow(self):
+        self._get_val_allow(-math.inf)
+
+    def test_get_Ininity_allow(self):
+        self._get_val_allow(math.inf)
+
```
Comment (rh-codebase) [inline test/test_dsa_store.py:128]: additional tests approved. add new tests as noted in other reviewed files.

File: dsautils/cli.py [hunk 26]
```diff
@@ -70,6 +70,47 @@ def radecel(mjd=None, localtime=None, utctime=None):
...
+        tu += 1000*int(tz.split(':')[0])*3600    # millisecond offset for time zone hours
+    elif mjd is None and localtime is None and utctime is not None:
+        assert utctime.count('-') == 2
+        tu = int(1000*Time(utctime, format='isot').unix)
+        tm = Time(utctime, format='isot').mjd
+    else:
+        print('Must provide either mjd or localtime')
+        return
+
+    query = f'SELECT time,airtemp FROM "wxmon" WHERE time >= {tu}ms and time < {tu+30000}ms'
```
[the remark this answers is not in the document]
Comment (caseyjlaw) [reply dsautils/cli.py:104]: That sounds like a nice way to remove duplicate code. But I'm confused: I have not needed to use authentication to get influx data.

File: dsautils/dsa_store.py (hunk 22 above)
Comment (caseyjlaw) [reply dsautils/dsa_store.py:364]: I don't think I made changes here.

File: dsautils/cli.py [hunk 27]
```diff
@@ -70,6 +70,47 @@ def radecel(mjd=None, localtime=None, utctime=None):
...
+    Time can be defined as mjd, local or UT time.
+    localtime string should have timezone attached to the string. utctime must not.
+    Unix date utility can get time from descriptive term, e.g.:
+    "> date --date='TZ="America/Los_Angeles" 10:00 yesterday' -Iseconds"
+    "2021-02-02T10:00:00-08:00"
+    The local time (with time zone) can be pasted into "localtime" argument to get values at that time.
+    """
+
+    if mjd is not None and localtime is None and utctime is None:
+        tu = int(1000*Time(mjd, format='mjd').unix)
```
[the remark this answers is not in the document]
Comment (caseyjlaw) [reply dsautils/cli.py]: Good point.

File: dsautils/cli.py [hunk 28]
```diff
@@ -70,6 +70,47 @@ def radecel(mjd=None, localtime=None, utctime=None):
...
+    query = f'SELECT time,airtemp FROM "wxmon" WHERE time >= {tu}ms and time < {tu+30000}ms'
+    try:
+        result = influx.query(query)
+        temp = float(result['wxmon']['airtemp'])
+        print(f'Temperature on {mjd}: {temp}C')
+
+    except KeyError:
+        print('No values returned by query.')
+
+
```
[the remark this answers is not in the document]
Comment (caseyjlaw) [reply dsautils/cli.py]: Sounds good.

File: dsautils/cli.py [hunk 29]
```diff
@@ -70,6 +70,47 @@ def radecel(mjd=None, localtime=None, utctime=None):
...
+    """ Get weather at any time.
+    Time can be defined as mjd, local or UT time.
+    localtime string should have timezone attached to the string. utctime must not.
+    Unix date utility can get time from descriptive term, e.g.:
+    "> date --date='TZ="America/Los_Angeles" 10:00 yesterday' -Iseconds"
+    "2021-02-02T10:00:00-08:00"
+    The local time (with time zone) can be pasted into "localtime" argument to get values at that time.
+    """
+
+    if mjd is not None and localtime is None and utctime is None:
```
[the remark this answers is not in the document]
Comment (caseyjlaw) [reply dsautils/cli.py:86]: I'll open an issue for this.

File: dsautils/cli.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (rh-codebase) [reply dsautils/cli.py:103]: Isn't this line:
influx = DataFrameClient('influxdbservice.sas.pvt', 8086, 'root', 'root', 'dsa110')
doing the authentication?  The 'root','root' part.

File: dsautils/dsa_store.py (hunk 22 above)
Comment (rh-codebase) [reply dsautils/dsa_store.py:364]: right. I'm reviewing my own changes which isn't really appropriate but I tried to be unbiased as possible. I'm making those changes and adding test cases.

File: dsautils/cli.py [hunk 30]
```diff
@@ -41,33 +41,32 @@ def radecel(mjd=None, localtime=None, utctime=None):
     """
 
     if mjd is not None and localtime is None and utctime is None:
-        tu = int(1000*Time(mjd, format='mjd').unix)
+        tu = int(MS_PER_SECOND*Time(mjd, format='mjd').unix)
```
Comment (rh-codebase) [inline dsautils/cli.py:44]: I should have spent a bit more time on this section. The nearly duplicated 3 lines of code in each if,else block only differ by the time and format parameter and could be converted into a function. Then each if block uses this function. The 1000 -> MS_PER_SECOND would have only needed to be replaced once.

File: dsautils/cli.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (caseyjlaw) [reply dsautils/cli.py:103]: Yep, that is the authentication!
I'll open an issue to restructure this. Good topic for a hack session.

**Comment by caseyjlaw** [comment]
Thanks, Rick! I didn't give you warning, but we agreed to assign PRs to a few of us during the OVRO/DSA visit. The goal is to practice a bit and learn each other's code.

**Pull request merged**